### PR TITLE
#74disagg

### PR DESCRIPTION
**Pull request by juliannosambatti**

### CODE REVIEW

File: pygypsy/disagg.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
+# -*- coding: utf-8 -*-
```
Comment (jo-tham) [inline pygypsy/disagg.py]: Lets remove this file from the history since it was changed to disaggregate.py

File: pygypsy/disaggregate.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
+# -*- coding: utf-8 -*-
+"""disaggregation of outpu_df into original species"""
+
+from __future__ import division
+
+import logging
+import pandas as pd
+
+original_file = pd.read_csv('/Users/juliannosambatti/Documents/Projects/gypsy/10-plots.csv')
```
Comment (jo-tham) [inline pygypsy/disaggregate.py]: i think we should avoid read data into global variables in a module here

do this in the unit test

also, those data are only on your local computer - can you subset them to one or two pertinent plots for the unit test and put them under `data/` here https://github.com/tesera/pygypsy/tree/dev/data

File: pygypsy/disaggregate.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+#    
+#    return df
+
+def smartway_to_build_disaggregated_simulated_df(df):
+
+    species = [('Aw', 'Aw'), ('Aw', 'Pb'), ('Sw', 'Fb'), ('Sw','Fd'), ('Sw', 'Sw')]
+    variables = ['bhage', 'tage', 'topHeight', 'BA', 'Gross_Total_Volume', 'Merchantable_Volume'] 
+
+    for gypsy_s, data_s in species:
+        df['SC_%s' %(data_s)] = df['SC_%s' %  (gypsy_s)] * df['PER_%s_data' % data_s] * df['prop_%s_class' %gypsy_s]
```
Comment (jo-tham) [inline pygypsy/disaggregate.py]: Why did you hoist these (SC and N_bh_--T) into the outer loop? it's a bit confusing to see a different approach for these variables than for the others

File: pygypsy/disaggregate.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+import logging
+import pandas as pd
+
+original_file = pd.read_csv('/Users/juliannosambatti/Documents/Projects/gypsy/10-plots.csv')
+simulated_data = pd.read_csv('/Users/juliannosambatti/Documents/Projects/gypsy/2012.csv')
+
+
+def sp_prop_total(perc_sp):
+    print perc_sp
+    total_prop = {'PER_Aw_data':0, 'PER_Pb_data':0, 'PER_Sw_data':0, 'PER_Fb_data':0, 'PER_Fd_data':0}
```
Comment (jo-tham) [inline pygypsy/disaggregate.py]: i think these would be more accurately called

`prop_--_data`

and

```
for sp, prop in perc_sp:
```

would be more accuratley called

```
for sp, perc in perc_sp:
```

File: pygypsy/disaggregate.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+def smartway_to_build_disaggregated_simulated_df(df):
+
+    species = [('Aw', 'Aw'), ('Aw', 'Pb'), ('Sw', 'Fb'), ('Sw','Fd'), ('Sw', 'Sw')]
+    variables = ['bhage', 'tage', 'topHeight', 'BA', 'Gross_Total_Volume', 'Merchantable_Volume'] 
+
+    for gypsy_s, data_s in species:
+        df['SC_%s' %(data_s)] = df['SC_%s' %  (gypsy_s)] * df['PER_%s_data' % data_s] * df['prop_%s_class' %gypsy_s]
+        df['N_bh_%sT' % (data_s)] = df['N_bh_%sT' % gypsy_s] * df['PER_%s_data' % data_s] * df['prop_%s_class' %gypsy_s]
+        for v in variables:
+            df['%s_%s' %(v, data_s)] = df['%s_%s' % (v, gypsy_s)] * df['PER_%s_data' % data_s] * df['prop_%s_class' %gypsy_s]
```
Comment (jo-tham) [inline pygypsy/disaggregate.py]: so we have 

new = gypsy * percent_species_in_data * proportion_gypsy_class

an example consider a plot with fd 10% and Sw 40% and Aw 50%

For gypsy, Sw is 50%
Fd is 20% of Sw

fd_density = gypsy_sw_density * percent_fd_data * proportion_fd_class
sw_density = gypsy_sw_density * percent_sw_data * proportion_sw_class

----------------------------

one thing that comes to mind here, is it really percent and proportion? if so i would expect

fd_density = 13.4 * 10 * .5

which is not correct. so mind the naming of the variables

-----------------------

an example consider a plot with fd 10% and Sw 40% and Aw 50%

fd_density = gypsy_sw_density * prop_fd_data * prop_sw_class
fd_density = 13.4 * .1 * .5

sw_density = gypsy_sw_density * prop_sw_data * prop_sw_class
sw_density = 13.4 * .4 * .5

-------------------------------------

that doesn't quite look right to me; am i missing something @juliannosambatti ?

if i'm not missing something, i think it should be 

fd_density = gypsy_sw_density * prop_fd_data / prop_sw_class
fd_density = 13.4 * .2

sw_density = gypsy_sw_density * prop_sw_data / prop_sw_class
sw_density = 13.4 * .8

File: tests/test_disagg.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+def test_sp_prop_total():
+    perc_sp = [
+                ('Aw', 50), ('Sw', 30), 
+                ('Pb', 20), ('', 0), 
+                ('', 0)
+               ]
+    
+    assert sp_prop_total(perc_sp) == {'PER_Aw_data':0.5, 'PER_Pb_data':0.2, 'PER_Sw_data':0.3, 'PER_Fb_data':0, 'PER_Fd_data':0}
+
+
```
Comment (jo-tham) [inline tests/test_disagg.py]: the best way to prove the routine works is to add a test here for the `build_disaggregated_simulated_df` function; you can use the forward simulation comparison tests or the data prep tests for reference - need to do that before merging this into dev anyway.

**Review by jo-tham — CHANGES_REQUESTED** [review]
Hey @juliannosambatti here's my review - i might have misunderstood some of the implementation; let me know if that's the case.

The main thing with respect to the correctness
 - i think it should be:

`disaggregatred_var_species = gypsy_var_species * proportion_data_species / proportion_gypsy_species`

note the swap of `*` with `/`

this is described in more detail in one of the comments

i also suggest to take the time to make sure everything is named well, this is a bit tricky because the items in question are hard to distinguish, but it's worth the time! a couple of the comments below relate to this (e.g. PER vs proportion)

**Comment by jo-tham** [comment]
@juliannosambatti a couple other things:

please fixup the tests which aren't passing; i can help if you're not sure how to resolve it - best thing to do is just post a comment on this pull request with an example of the test output indicating why its not passing if you get stuck

------------------------

for future reference please keep commit messages brief, ~50 characters

There is a tesera standard for this borrowed from the git community:
https://github.com/tesera/knowledge/wiki/One-Convention#the-seven-rules-of-a-great-git-commit-message

if you have more details to add and you are using the CLI you can use

```
git commit -m 'this is the summary of what was done

here are more detailed
stuff about the commit
and such'
```

**Comment by juliannosambatti** [comment]
You are right up there. The main problem is to avoid a division by zero when proportion of gypsy speceis is zero.

**Comment by jo-tham** [comment]
@juliannosambatti use this

``` python
try:
    # the division
except ZeroDivisionError:
    # what to do if the gypsy species is 0; probably just assign the variable to 0
```

**Comment by juliannosambatti** [comment]
Actually this is wrong too. The class proportion is the sum of component proportion. Thus we just need to multiply by the original proportion to get the species variable

**Comment by jo-tham** [comment]
> Actually this is wrong too. The class proportion is the sum of component proportion. Thus we just need to multiply by the original proportion to get the species variable

@juliannosambatti can you make an example in comment here?

**Review by jo-tham — APPROVED** [review]

**Pull request merged**